feat: Add model loading using `tensorizer`

sangstar opened the pull request:
# Tensorizer Support
This PR allows models used for the OpenAI-compatible API server to be loaded using
[Coreweave's Tensorizer](https://github.com/coreweave/tensorizer), enabling extremely fast model loads from HTTP/HTTPS, 
Redis, and S3 endpoints. 

The key changes involved are:

1. Listing `tensorizer>=2.8.1` as a runtime dependency in `requirements.txt`
2. Adds `tensorizer_loader.py` to `vllm/model_executor` that provides utility functions
for tensorizer. 
3. Adds multiple args to the vLLM's OpenAI inference service entrypoint that allows
the user to specify the path to serialized-by-tensorizer model tensors, as well as 
arguments for tensorizer's deserializer. 
4. Allows deserialization of serialized model tensors in HuggingFace's model format, 
as well as supporting deserializing serialized vLLM-formatted models, allowing the 
use of loading with `plaid_mode`, which can allow Llama 2 13B to start serving requests
in as little as 10 seconds. Also supports encrypting and decrypting model tensors.
5. Adds a `tensorize_vllm_model.py` script to `examples/` that allows vLLM models to be serialized and
deserialized with `tensorizer`, as well as with supporting encryption.

Credentialing for S3 is supported by passing a user's access and secret key to 
`S3_ACCESS_KEY_ID` and `S3_SECRET_ACCESS_KEY` environment variables respectively

# Model loading benchmarks

Tensorizer can load models like Llama 2 13B in nearly 10 seconds. In order to do so, a model must be 
serialized using `TensorSerializer` to a `.tensors` file located either locally or through a S3, HTTP/HTTPS, or Redis
endpoint. `--tensorizer-uri` must be specified with the serialized tensors location when invoking the API server. 

Example usage:

```
python -m vllm.entrypoints.openai.api_server \
--host 0.0.0.0 \
--model EleutherAI/pythia-6.9b \
--load-format tensorizer \
--tensorizer-uri s3://tensorized/EleutherAI/pythia-6.9b/fp16/model.tensors

```

If a vLLM model is serialized, `plaid_mode` can be used, which loads much faster. The following plot demonstrates model loading time benchmarks for vLLM's OpenAI-compatible inference server on a Nvidia A40 GPU.

![torch-extras (3)](https://github.com/coreweave/vllm/assets/51196172/4494012f-b5b5-4ee0-805b-5f0a01f175e5)

sangstar committed:
feat: Support loading model tensors using `tensorizer`

This feature allows vLLM models to be loaded extremely fast using
`tensorizer`. `tensorizer` loads serialized model tensors from
HTTP/HTTPS, Redis, S3 endpoints, or locally, typically on the scale
of multiple GB/s.

sangstar committed:
fix: Remove unnecessary files

sangstar committed:
fix(vllm-tensorizer): Allow providing S3 credentials

This allows the deserializer to access S3 credentials when reading.
It extracts the access key from env variables `S3_ACCESS_KEY_ID`
and secret key from `S3_SECRET_ACCESS_KEY`.

sangstar committed:
fix: Fix passing S3 auth vars through stream

Previous commit wasn't able to pass S3
credentials through `TensorDeserializer`, so
it is instead passed to `stream_io.openstream`
which is used to instantiate the
`TensorDeserializer`.

sangstar committed:
fix: Disallowing `plaid_mode = False` and updating `tensorizer` version

Removed functionality to allow for tensorizing without `plaid_mode`,
and updating to `tensorizer==2.8.0`

sangstar committed:
refactor: Retire use of `download_dir` as `TensorizerArgs` param

Replaces `download_dir` with `tensorizer_uri` as a `TensorizerArgs`
param. This is due to discussions on `download_dir` being a confusing
and ultimately not helpful parameter for the location of model tensors.
Instead, `download_dir` is back to the definition coinciding with
the convention set by HuggingFace; as a location to download weights
for caching. A new parameter takes its place: `tensorizer_uri`. This
specifically deals with locating model tensors for `tensorizer`.

sangstar committed:
fix: Remove `store_true` action for `--tensorizer-uri`

Also slight formatting fix for warning when loading weights with
`download_dir` not set to `None`.

sangstar committed:
refactor: No 2x copying for `tensorizer` (WIP)

sangstar committed:
chore: Omit commandeering weight loaders for merging layers (WIP)

sangstar committed:
feat: Re-add deserializing vLLM models

Integrated changes from
ssteel/tensorizer-support branch that allowed
for deserializing vLLM models.

sangstar committed:
chore: Harmonize CPU and GPU deserializing

Integrated previous support for vLLM-formatted
model loading with `tensorizer` that makes full
use of loading to the GPU with `plaid-mode`, as
well as falling back on being able to load
HuggingFace models for serving using the CPU so
that vLLM can perform its manual GPU loading.

sangstar committed:
perf: Add `force_http=True` for faster loading speeds

sangstar committed:
chore: Reformat code with `format.sh`, cleanup debugging code

sangstar committed:
chore: Fix formatting, some misc. changes

Fixed some unnecessary formatting changes in `arg_utils.py`,
`weight_utils.py` and `model_loader.py` fixed improperly passing
`force_http` to `TensorDeserializer` rather than `open_stream`.

wbrown commented on requirements.txt:
```diff
@@ -4,6 +4,7 @@ ray >= 2.9
 sentencepiece  # Required for LLaMA tokenizer.
 numpy
 torch == 2.1.2
+tensorizer == 2.8.0
```
Do we want this to be fixed on `2.8.0`?

dmarx replied:
recommend changing it to `>=`

sangstar replied:
Fine with this if we don't expect any forward compatibility issues (which I
imagine is very much not an issue)

wbrown commented on vllm/config.py:
```diff
@@ -79,6 +79,7 @@ def __init__(
         quantization: Optional[str] = None,
         enforce_eager: bool = False,
         max_context_len_to_capture: Optional[int] = None,
+        tensorizer_args=None,
```
Formatting does not match. You also do not type this.

dmarx replied:
> You also do not type this.
to clarify: `tensorizer_args` needs a type hint

wbrown commented on vllm/engine/arg_utils.py:
```diff
@@ -294,7 +299,14 @@ def from_cli_args(cls, args: argparse.Namespace) -> 'EngineArgs':
-        engine_args = cls(**{attr: getattr(args, attr) for attr in attrs})
+        engine_args = cls(**{
+            attr: getattr(args, attr)
+            for attr in attrs if hasattr(args, attr)
```
Maybe document in a comment why we do this.

wbrown commented on vllm/engine/arg_utils.py:
```diff
@@ -307,7 +319,7 @@ def create_engine_configs(
             self.dtype, self.seed, self.revision, self.code_revision,
             self.tokenizer_revision, self.max_model_len, self.quantization,
-            self.enforce_eager, self.max_context_len_to_capture,
+            self.enforce_eager, self.max_context_len_to_capture, self.tensorizer_args,
```
Insert a newline here, and combine with following line. You're bleeding over the columns.

wbrown commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,184 @@
+
+
+@dataclass
+class TensorizerArgs:
```
Docstring this please.

wbrown commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,184 @@
+@dataclass
+class TensorizerArgs:
+    tensorizer_uri: Union[io.BufferedIOBase, io.RawIOBase, typing.BinaryIO,
+                          str, bytes, os.PathLike, int, ]
```
Remove trailing `,`.

wbrown commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,184 @@
+
+
+class TensorizerAgent:
+
```
Remove newline here, or insert a docstring here. Preferably a docstring.

wbrown commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,184 @@
+                f"file when load_format = tensorizer")
+
+    def deserialize(self):
+        before_mem = get_mem_usage()
```
Docstring.

sangstar committed:
fix: Correct logging for loading tensorizer with cpu

wbrown commented on vllm/model_executor/weight_utils.py:
```diff
@@ -134,7 +136,8 @@ def prepare_hf_model_weights(
 ) -> Tuple[str, List[str], bool]:
     # Download model weights from huggingface.
-    is_local = os.path.isdir(model_name_or_path)
+    is_local = os.path.isdir(
```
Escape with `\` after `)`
```
... isdir(model_name_or_path) \
       and load_format != "tensorizer"
```

dmarx replied:
Alternatively, if you encapsulate the expression in parens you don't need the line continuation `\`, i.e. 

```python
is_local = (os.path.isdir(model_name_or-path)
    and load_format != "tensorizer")
```

also this is more of a stylistic thing so take it or leave it, but whenever I have a multi-expression conditional like this I generally try to encapsulate each subexpression in parens to mitigate order of operations ambiguity. So not a required change, but if you wanna be extra explicit you could do:

```python
is_local = (os.path.isdir(model_name_or-path)
    and (load_format != "tensorizer"))
```

sangstar replied:
I like both suggestions! I'll implement this

wbrown reviewed (changes requested):
Graph is confusing. I'd just combine the two graphs and then have three bars for each.

Formatting/documentation changes requested, others may have more insightful comments.

dmarx commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,16 @@
+# model_ref = "EleutherAI/gpt-neo-125M"
+model_name = model_ref.split("/")[-1]
+# Change this to your S3 bucket.
+s3_bucket = "bucket"
```
recommend using the `tensorized` bucked here and dropping the `model_name = model_ref.split("/")[-1]` bit so the example works unmodified out of the box

sangstar replied:
This is serializing though -- no one should be writing to `tensorized`.

sangstar replied:
I'm going to refactor this though

dmarx commented on vllm/model_executor/weight_utils.py:
```diff
@@ -250,6 +262,20 @@ def hf_model_weights_iterator(
                 param = np.load(f)
             yield name, torch.from_numpy(param)
+    elif load_format == "tensorizer":
+        logger.warning("Deserializing HuggingFace models are not optimized for "
```
pretty sure "are not optimized" should be "is not optimized". The subject of the verb here is "(tensorizer) deserializing"

dmarx commented on vllm/model_executor/weight_utils.py:
```diff
@@ -250,6 +262,20 @@ def hf_model_weights_iterator(
+        logger.warning("Deserializing HuggingFace models are not optimized for "
+                       "loading on vLLM, as tensorizer is forced to load to CPU. "
+                       "Consider deserializing a vLLM model instead for faster "
+                       "load times. See the `tensorize_vllm_model.py` example "
```
reference the actual path to the script here, so `tensorize_vllm_model.py` -> `examples/tensorize_vllm_model.py`

sangstar committed:
chore: Implement changes from feedback

Misc. fixes from now resolved conversations. Mostly consisting of
changes to syntax, style, adding docstrings, and versioning.

sangstar commented:
> Graph is confusing. I'd just combine the two graphs and then have three bars for each.
> 
> Formatting/documentation changes requested, others may have more insightful comments.

I only have `plaid_mode` comparisons for Llama 2 13B, but I can have 3 bars for the Llama 2 13B in that case

sangstar committed:
fix: Correctly instantiate vLLM-formatted models

`examples/tensorize_vllm_model.py` now correctly instantiates
vLLM-formatted models.

sangstar committed:
chore: Reformat and delete deprecated comment from `.ipynb`

sangstar committed:
perf: Allow passing of deserializer args from `TensorizerArgs`

sangstar committed:
style: Reformat with new formatting changes

Eta0 commented on vllm/engine/arg_utils.py:
```diff
@@ -123,7 +128,10 @@ def add_cli_args(
             '"dummy" will initialize the weights with random values, '
-            'which is mainly for profiling.')
+            'which is mainly for profiling.'
+            '"tensorizer" will load the weights using tensorizer from CoreWeave,'
```
Missing a space between these two sentences.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,137 @@
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Arguments for serializing and deserializing models. "
+        "in this example script.")
```
This seems to be an unfinished sentence. `ArgumentParser` descriptions should describe the entire program.

sangstar committed:
Run yapf and ruff

Refined formatting to adhere to vLLM's specifications in `format.sh`
Commit message neglecting conventional commit templating due to
vLLM's preference for commit messaging on using their formatter.

sangstar committed:
fix: Fix incorrect `TensorizerArgs` import in `config.py`

Also enforces a default float16 precision for models when serializing

Eta0 commented on requirements.txt:
```diff
@@ -4,6 +4,7 @@ ray >= 2.9
 sentencepiece  # Required for LLaMA tokenizer.
 numpy
 torch == 2.1.2
+tensorizer >= 2.8.0
```
Force people to update their tensorizer versions more! (When there are no breaking changes)
```suggestion
tensorizer >= 2.8.1
```

Eta0 commented on vllm/engine/arg_utils.py:
```diff
@@ -294,7 +303,20 @@ def from_cli_args(cls, args: argparse.Namespace) -> 'EngineArgs':
+            # Create an instance of TensorizerArgs using the
+            # from_cli_args method
+            engine_args.tensorizer_args = TensorizerArgs.from_cli_args(args)
         return engine_args
```
The change to `if hasattr(args, attr)` doesn't do what the comment implies. The original already plucked out only the parts of `args` that were present in `attrs` (i.e., the relevant `EngineArgs` fields), and would leave any unknown ones untouched. That is to say, this slight extension of the original code suffices (plus using `vars` for succinct-ness):

```py
@classmethod
def from_cli_args(cls, args: argparse.Namespace) -> 'EngineArgs':
    # Get the list of attributes of this dataclass.
    attrs = [attr.name for attr in dataclasses.fields(cls)]
    opts: dict = vars(args)

    # Instantiate engine_args with the subset of CLI args pertaining to it.
    # There may be other CLI args that pertain only to tensorizer.
    engine_args = cls(**{attr: opts[attr] for attr in attrs})

    # Check if tensorizer_args are expected.
    if args.load_format == "tensorizer":
        tensorizer_attrs = [
            attr.name for attr in dataclasses.fields(TensorizerArgs)
        ]
        # Create an instance of TensorizerArgs with its subset of the CLI args.
        engine_args.tensorizer_args = TensorizerArgs(
            **{attr: opts[attr] for attr in tensorizer_attrs}
        )

    return engine_args
```

The `if hasattr(args, attr)` change actually makes it so that you can supply an incomplete set of `EngineArgs` constructor parameters or `TensorizerArgs` constructor parameters (if the constructors cooperate and supply default values for missing parts).

If you want that functionality, it can also be written more cleanly (and robustly):

```py
@classmethod
def from_cli_args(cls, args: argparse.Namespace) -> 'EngineArgs':
    # Get the attributes of this dataclass.
    attrs = {attr.name for attr in dataclasses.fields(cls)}
    opts: dict = vars(args)

    # Instantiate engine_args with the subset of CLI args pertaining to it.
    engine_keys = opts.keys() & attrs
    engine_args = cls(**{k: opts[k] for k in engine_keys})

    # There may be other CLI args that pertain only to tensorizer.
    other_keys = opts.keys() - attrs

    # Check if tensorizer_args are expected.
    if args.load_format == "tensorizer":
        # Create an instance of TensorizerArgs from the remaining CLI args.
        engine_args.tensorizer_args = TensorizerArgs(
            **{k: opts[k] for k in other_keys}
        )
    elif other_keys:
        raise ValueError(f"Unused CLI arguments: {sorted(other_keys)}")

    return engine_args
```

This would ensure that any arguments eaten by `EngineArgs` are not passed on to `TensorizerArgs`, to avoid conflicts between the two. The `ValueError` part is optional and would tell you if there were extraneous args that the code isn't using, if you expect to use them all. If you expect there to be extraneous args you would instead pluck out the `TensorizerArgs` attrs explicitly, but IMO even that will be a lot neater using `opts.keys() & tensorizer_attrs` than an opaque comprehension full of `getattr` and `hasattr`.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+def _make_model_contiguous(model: nn.Module):
+    # Ensure tensors are saved in memory contiguously
+    for param in model.parameters():
+        param.data = param.data.contiguous()
```
Why is this here?

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+        param.data = param.data.contiguous()
+
+
+#
```
```suggestion
```

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+            "s3_access_key_id": self.s3_access_key_id,
+            "s3_secret_access_key": self.s3_secret_access_key,
+            "s3_endpoint": self.s3_endpoint,
+            "force_http": True,
```
It is not acceptable to enable insecure HTTP unconditionally. This must be opt-in, because unless the user is sure that their data is already being transferred over a secured, private network (or they are only using public models, or have set up tensorizer encryption), this will leak model weights.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+            "force_http": True,
+        }
+
+        # Omitting self.dtype and self.device as this behaves weirdly
```
Could you elaborate? (And remove these parameters from `TensorizerArgs` if they are always ignored).

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+    plaid_mode_buffers: Optional[int] = None
+    verify_hash: bool = False
+    filter_func: Optional[Callable[[str], Union[bool, Any]]] = None
+    deserializer_encryption_key: Optional[str] = None
```
`deserializer_encryption_key` is either a binary key (more secure if generated correctly) or an arbitrary string (usually slower and less secure). They are differentiated by the `DecryptionParams` factory functions. Only accepting a string here means models encrypted with a binary key can't be loaded. Disambiguating them automatically is, in general, not possible, so the encryption type needs to be specified by the user.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+        self.deserializer_params = {
+            "filter_func": self.filter_func,
+            "lazy_load": self.lazy_load,
+            "plaid_mode": bool(_is_vllm_model(file_uri=self.file_obj)),
```
This choice doesn't make sense. If the intent is to deserialize non-vLLM models to the CPU, `plaid_mode` does not control the output device; if you try to enable it while deserializing to the CPU it just throws an error. `plaid_mode` should always be enabled if available, which is the default behaviour of leaving `plaid_mode=None`.

sangstar replied:
Point taken -- will decouple `plaid_mode` from any logic worrying about whether the model is a vLLM or HF model or not

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+            action="store_true",
+            help="An `EncryptionParams` object holding a password or key"
+            "to use for encryption. If None, no encryption will be used.",
+        )
```
`"store_true"` (so, on-or-off) doesn't make sense here given the help text. Also, the help text is missing a space between "key" and "to."

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+            default=None,
+            help="A `DecryptionParams` object holding a password or key to"
+            "use for decryption. ``None`` (the default) means no decryption.",
+        )
```
Like the encryption during serialization, this cannot be adequately specified with a single string option. For the correct way to expose tensorizer encryption via CLI, [refer to my tensorizer encryption CLI tool in the tensorizer repo](https://github.com/coreweave/tensorizer/blob/v2.8.1/examples/encrypt_existing.py).

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+    def _init_model(self):
+        model_args = self.model_config.hf_config
+        model_args.torch_dtype = self.model_config.dtype
+        model = no_init_or_tensor(lambda: self.model_cls(*[model_args]))
+        return model
```
`*[model_args]` is equivalent to `model_args`. Did you mean `*model_args` (no `[]`)? Other than that, `no_init_or_tensor` is better used as a context manager:
```suggestion
        with no_init_or_tensor():
            return self.model_cls(*model_args)
```

sangstar replied:
`model_args` isn't an iterable, but `model_cls` expects a `Config` in its `__init__`, so just `model_args` is appropriate

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+        if not self.tensorizer_args.tensorizer_uri.endswith(".tensors"):
+            raise ValueError(
+                f"download_dir {self.tensorizer_args.tensorizer_uri} must "
+                f"specify a .tensors file when load_format = tensorizer")
```
Is this used anywhere? A filename check seems unnecessary; tensorizer will already raise an error if it detects it is given an incorrect filetype based on the file contents rather than the name. This would just make it harder to use stuff like temporary files with throwaway names.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+            stream, **self.tensorizer_args.deserializer_params)
+        deserializer.load_into_module(self.model)
+        self.model = self.model.to(dtype=self.model_config.dtype)
+        end = time.time()
```
1. This isn't necessarily lazy-loading or S3
2. Use a `TensorDeserializer` as a context manager for better cleanup
3. Use `time.perf_counter()` for better timing
```suggestion
        start = time.perf_counter()
        with stream_io.open_stream(
            self.tensorizer_args.tensorizer_uri,
            mode="rb",
            **self.tensorizer_args.stream_params,
        ) as stream, TensorDeserializer(
            stream, **self.tensorizer_args.deserializer_params
        ) as deserializer:
            deserializer.load_into_module(self.model)
            self.model = self.model.to(dtype=self.model_config.dtype)
            end = time.perf_counter()
```
Using `.to(dtype=...)` instead of tensorizer's default `dtype` support is also confusing; why is that done?

Eta0 commented on vllm/model_executor/weight_utils.py:
```diff
@@ -250,6 +262,21 @@ def hf_model_weights_iterator(
+            "loading on vLLM, as tensorizer is forced to load to CPU. "
+            "Consider deserializing a vLLM model instead for faster "
+            "load times. See the examples/tensorize_vllm_model.py example "
+            "script for serializing vLLM models.")
```
Consider using Python's [built-in warning support](https://docs.python.org/3/library/warnings.html) for this since it has more filter options in user code. We should also add an environment variable (or other toggle) to silence this completely.

Eta0 commented on vllm/model_executor/model_loader.py:
```diff
@@ -83,6 +89,15 @@ def get_model(model_config: ModelConfig, device_config: DeviceConfig,
             # Load the weights from the cached or downloaded files.
-            model.load_weights(model_config.model, model_config.download_dir,
-                               model_config.load_format, model_config.revision)
+            if model_config.load_format == "tensorizer":
+                # Provide a dynamic load format for `model.load_weights`
+                # to retain tensorizer args from CLI.
+                model_config.load_format = ("tensorizer",
+                                            model_config.tensorizer_args)
+            model.load_weights(
+                model_config.model,
+                model_config.download_dir,
+                model_config.load_format,
+                model_config.revision,
+            )
```
This is a sort of bizarre way to pack this into the parameters. It would fit better as an optional argument for `load_weights`, maybe passed around in `**kwargs` for generality's sake. If you need to pack it into `load_format` for some reason, you could make it a `str` subclass with extra information so that all other code can continue to treat it as a string, e.g.:
```py
class ParameterizedLoadFormat(str):
    __slots__ = "params"

# ...Later...
model_config.load_format = ParameterizedLoadFormat(model_config.load_format)
model_config.load_format.params = model_config.tensorizer_args

# Can still be used as a string for most purposes
print(model_config.load_format)  # Prints "tensorizer"
assert model_config.load_format == "tensorizer"  # Passes
```
So that all code from here on doesn't have to handle both cases of the format being either `tuple` or `str` just for tensorizer's sake. This wouldn't make it much less of a strange way to pass this data around, but it would be at least be unobtrusive, which is what I suppose the motivation was for not adding an optional argument here to begin with.

sangstar replied:
This strange workaround was predicated on the fact that `model.load_weights` is defined for every single model class in `vllm/model_executor/models/*.py` 

I sorely wish that their models inherited from some base class because there's seemingly a lot of copy-pasting for their model class methods, but I felt this was the least intrusive way of making this change, rather than adding this option to every single model class and forcing this extra arg to be passed as new models are added, as this is technically a work-around that applies for all model classes extensibly within a few awkward lines of code. 

I very much like your suggestion though so I'll go with that.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+@dataclass
+class TensorizerArgs:
+    tensorizer_uri: Union[io.BufferedIOBase, io.RawIOBase, typing.BinaryIO,
+                          str, bytes, os.PathLike, int, ]
```
```suggestion
    tensorizer_uri: Union[io.BufferedIOBase, io.RawIOBase, typing.BinaryIO,
                          str, bytes, os.PathLike, int]
```

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+    filter_func: Optional[Callable[[str], Union[bool, Any]]] = None
+    deserializer_encryption_key: Optional[str] = None
+    """
+  Args for TensorizerAgent class. These are used to configure the behavior of 
```
```suggestion
  Args for the TensorizerAgent class. These are used to configure the behavior of 
```

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+            help="If True, tensors will be loaded and cached when keys are"
+            "accessed. If False, all tensors will be loaded into memory up"
+            "front.",
+        )
```
If vLLM's access pattern will not use this option effectively anyway (e.g. always requiring every tensor to be loaded), it shouldn't be exposed as a configurable option.

sangstar replied:
Agreed.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+            help="The number of buffers to use in plaid mode."
+            "This is only used if ``plaid_mode=True``. These buffers"
+            "are used to pipeline the loading and processing of tensors.",
+        )
```
This is an (optional) `int`, includes details that don't make sense with the current CLI, and is missing spaces.

```suggestion
        parser.add_argument(
            "--plaid-mode-buffers",
            type=int,
            default=None,
            help="The number of buffers to use to pipeline"
            " the loading and processing of tensors.",
        )
```

Alternatively, we could skip including it, as this option is about to be rendered obsolete (and deprecated) in the next release of tensorizer.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+        )
+        parser.add_argument(
+            "--tensorizer-uri",
+            help="Path to serialized model tensors. Can be a local file path"
+            "or a S3 URI.",
```
```suggestion
            help="Path to serialized model tensors. Can be a local file path,"
            " or an HTTP(S) or S3 URI.",
```

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+        parser.add_argument(
+            "--verify-hash",
+            action="store_true",
+            help="If True, the hashes of each tensor will be verified"
+            "against the hashes stored in the metadata. A `HashMismatchError`"
+            "will be raised if any of the hashes do not match.",
```
Missing spaces, and is written for a programmatic interface rather than a CLI.
```suggestion
            help="If enabled, the hashes of each tensor will be verified"
            " against the hashes stored in the file metadata. An exception"
            " will be raised if any of the hashes do not match.",
```

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,233 @@
+    @staticmethod
+    def add_cli_args(
+        parser: argparse.ArgumentParser, ) -> argparse.ArgumentParser:
+        """Tensorizer CLI arguments"""
```
Since these options are only valid for tensorizer, either their help text or layout in a CLI menu needs to communicate that requirement. This would be best represented by putting them in a tensorizer-specific [`argparse` argument group](https://docs.python.org/3/library/argparse.html#argument-groups).

sangstar committed:
perf: Multiple misc. improvements from code review

Multiple fixes and performance improvements from code reviews.
Some fixes include adjusting the `tensorizer` version, an overhaul
of tensorize_vllm_model.py to allow for encryption as well as other
improvements, and some slight fixes to tensorizer_loader.py

sangstar committed:
pref: More misc. fixes to complete initial code review

A multitude of fixes requested from code reviewers have been
implemented, ranging from encryption support, more logical handling
of tensorizer parameters and argparsing, documentation syntax fixes,
and some function fixes

sangstar commented on vllm/engine/arg_utils.py:
```diff
@@ -300,23 +300,20 @@ def add_cli_args(
+
         # Set the attributes from the parsed arguments.
+        engine_args = cls(**{attr: getattr(args, attr) for attr in attrs})
 
```
As per previous sticking point, I initialize `tensorizer_args` as `None` and aim to populate it here and return that adjusted `engine_args`, which is achieved here. This was the original aim which is accomplished here without being prone to the previous mistake (to my knowledge?)

Eta0 replied:
In this code, any overlap between the tensorizer arguments and engine arguments will be given to both classes, and any arguments unused by either will be ignored without any warning. If that is acceptable behaviour in this case, then this change ought to work. Though, it would still be cleaner using a dict instead of copious `getattr` and `hasattr` on namespaces.

sangstar replied:
I've decided to change it to this:

```
    @classmethod
    def from_cli_args(cls, args: argparse.Namespace) -> 'EngineArgs':
        # Get the list of attributes of this dataclass.
        attrs = [attr.name for attr in dataclasses.fields(cls)]
        # Set the attributes from the parsed arguments.
        engine_args = cls(**{attr: getattr(args, attr) for attr in attrs})

        engine_args.tensorizer_args = TensorizerArgs.from_cli_args(args) if (
                engine_args.load_format == "tensorizer") else None

        return engine_args
```

Which is the least intrusive case, as it just adds the line initializing `engine_args.tensorizer_args`, and I've instead made `tensorizer_args` a *dynamic* attribute for `engine_args`, which I think is further less intrusive because I'm not adding a new instance attribute for `engine_args`. 

I'm aware that this hands off the entire list of `args` to `TensorizerArgs`, but `TensorizerArgs.from_cli_args` expects this and parses accordingly.

sangstar commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -147,14 +185,8 @@ def __init__(
-                f"download_dir {self.tensorizer_args.tensorizer_uri} must "
-                f"specify a .tensors file when load_format = tensorizer")
+        with no_init_or_tensor():
+            return self.model_cls(model_args)
```
`model_args` is not an iterable, but instead a `Config` file. This should be appropriate and has been tested.

sangstar committed:
fix: Remove `print(tensorizer_args)`

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
-    ## TODO: Decide if below two are worth including besides --model and
-    ## --tensorizer-uri
-    parser = AsyncEngineArgs.add_cli_args(parser)
+        description="Arguments for serializing and deserializing vLLM "
```
`ArgumentParser` descriptions are for the program, not just its arguments, so beginning with "arguments for [...]" is inappropriate. It becomes the intro paragraph to the help text for the program, right after the usage syntax line.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
+        "--model",
+        type=str,
+        required=True,
+        help="The model reference name to serialize or deserialize"
```
We should elaborate on what formats are supported for the `"model reference name"`. Only HuggingFace IDs? Paths? (To specific files? Folders? On the local filesystem?). Maybe provide an example value here as well.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
+             "previous example, the --input-dir to use would be "
+             "s3://BUCKET/EleutherAI/pythia-1.4b/HASH"
+
+    )
```
1. Extended advice about object storage buckets versus local directories can be moved to the `epilog` of the `ArgumentParser` for better readability and discoverability
2. "`for this reason, --input-dir is recommended to specify a bucket name if using object storage rather than a local dir`" — in what case would someone specify a local directory if using object storage? Doesn't that mean they aren't using object storage?
  a. The recommendation worded as *"for this reason"* doesn't really specify what advantage it has over the alternative
4. The name `--input-dir` is confusing when it is also an output. You can just call it `--directory`, or `--serialized-directory`, or something.
  a. It's further confusing that the same parameter requires different values during serialization and deserialization (the latter requiring you to find the hash and append it to the command, the former being a prefix).
  b. Both of the above points could be addressed by splitting these into different parameters (possibly under their respective subparsers), each with only their relevant part of this help text.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
+        required=False,
+        help=("File to write 32 bytes of randomly-generated binary data used as"
+              " an encryption key"
+              )
```
Since this enables encryption, that should be clear in the description. Also, the number of key bytes isn't necessary to assert here.
```suggestion
        help=(
            "Encrypt the model weights with a randomly-generated binary key,"
            " and save the key at this path"
        )
```

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
+        type=str,
+        required=True,
+        help="Decryption keyfile to use to decrypt the model"
+    )
```
`required` shouldn't be `True` if encryption is optional, and the description can be a little more precise.
```suggestion
    deserialize_parser.add_argument(
        "--keyfile",
        type=str,
        required=False,
        help=(
            "Path to a binary key to use to decrypt the model weights,"
            " if the model was serialized with encryption"
        )
    )
```

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -56,60 +124,36 @@ def _get_model_architecture(config: PretrainedConfig) -> Type[nn.Module]:
-initialize_model_parallel()
-
-
 def serialize():
```
Parameters are better than globals

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -56,60 +124,36 @@ def _get_model_architecture(config: PretrainedConfig) -> Type[nn.Module]:
+    if keyfile:
+        with _write_stream(keyfile_path) as stream:
+            stream.write(encryption_params.key)
+    print("Is encryption being used?", encryption_params is not None)
```
Leftover debug print statement
```suggestion
```

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
+             "directory under a directory given by the model ref set "
+             "in --model and a unique identifier. For instance, if "
+             "serializing EleutherAI/pythia-1.4b, in S3 bucket "
+             "BUCKET, tensors and encryption key if applicable will "
```
The encryption key's path should always be specified separately. One of the benefits of encrypted models is that they can be put in public object storage buckets, and automatically putting the encryption key right next to the encrypted weights in one of those would be a... questionable decision.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
+        'deserialize',
+        help=(
+            "Deserialize a model from `--input-dir`"
+            " to verify it can be loaded and used."
```
No period here, for consistency
```suggestion
            " to verify it can be loaded and used"
```

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
+        "--dtype",
+        type=str,
+        required=False,
+        help="The dtype to use for the model"
```
The default behaviour here should be noted (cast everything vs. use the originals?)

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -4,38 +4,106 @@
-## TODO: Generally clearn this up
 
 def parse_args():
     parser = argparse.ArgumentParser(
```
It should be noted somewhere that making use of encryption/decryption support requires the `libsodium` library to be installed as an optional dependency.

sangstar replied:
Hopefully this updated program description is satisfactory:

```
        description="An example script that can be used to serialize and "
                    "deserialize vLLM models. These models "
                    "can be loaded using tensorizer directly to the GPU "
                    "extremely quickly. Tensor encryption and decryption is "
                    "also supported, although libsodium must be installed to "
                    "use it.")
```

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -143,8 +191,35 @@ def deserialize():
+
+keyfile = args.keyfile if args.keyfile else None
+
+base_path = f"{input_dir}/{model_ref}/{uuid.uuid4().hex}" \
```
Since a UUID is not a hash this will even serialize identical data to different locations on each run. That is not necessarily a bad thing, but it's worth considering.

This randomized component additionally gets in the way of automated processing, since the user needs to query their storage (e.g. send an authenticated request to list objects in an object storage bucket) and parse the results to find where their model went. This script should both:
1. Output where the object is being saved (in some easy-to-parse way), and
2. Allow setting a custom suffix instead of a random one, for workflows that require it.

sangstar replied:
I'll propose a `suffix` argument that can be given to a `serialize` command, and if none is given it'll take the UUID. It'll also spit out the location of the saved tensors in stdout.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -56,60 +124,36 @@ def _get_model_architecture(config: PretrainedConfig) -> Type[nn.Module]:
     to_dtype = _get_and_verify_dtype(config, dtype=dtype)
     model = model.to(dtype=to_dtype)
-    print("model precision is now", {next(model.parameters()).dtype})
     model.save_pretrained(DOWNLOAD_DIR)
```
Does this really have to be saved to disk? It should at least be cleaned up afterwards.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -143,8 +191,35 @@ def deserialize():
+model_name = model_ref.split("/")[1]
+input_dir = args.input_dir.rstrip('/')
+
+DOWNLOAD_DIR = f"/tmp/{model_name}"
```
Use the `tempfile` module for this.

Eta0 commented on examples/tensorize_vllm_model.py:
```diff
@@ -143,8 +191,35 @@ def deserialize():
+
+DOWNLOAD_DIR = f"/tmp/{model_name}"
+
+os.environ["MASTER_ADDR"] = "0.0.0.0"
+os.environ["MASTER_PORT"] = "8080"
+
+torch.distributed.init_process_group(world_size=1, rank=0)
```
Is this necessary for a single worker? The `MASTER_ADDR` should probably be set to loopback, at least.

sangstar replied:
vLLM throws an assertion error if there isn't parallelism when loading a vLLM model:

```
AssertionError: tensor model parallel group is not initialized
```

So I figured this would be the path of least resistance. I'll change the `"MASTER_ADDR"` to loopback though.

Eta0 commented on vllm/engine/arg_utils.py:
```diff
@@ -300,23 +300,20 @@ def add_cli_args(
+        # TensorizerArgs.
+        engine_attrs = [attr.name for attr in dataclasses.fields(cls)
+                        if attr.name != 'tensorizer_args']
+        attrs = [attr for attr in engine_attrs]
```
Please reread this line of code...

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -33,49 +31,50 @@ def _is_vllm_model(model_config: ModelConfig = None,
+          no decryption. The file must be created a priori using 
+          DecryptionParams. See the example script in 
+          examples/tensorize_vllm_model.py. A relative path within the 
+          tensorizer_uri directory can also be used, or an absolute path.
```
- "`encryption-keyfile`" → `encryption_keyfile`.
- "`containing a password or key`" — Using a password is not (currently) an option in this setup (though it is in tensorizer), this is only for a binary key.
- "`The file must be created a priori`" — This is not an appropriate use of *a priori*. If you mean "in advance," you can say that, but in general using the contents of a file comes after creating it, so this is sort of needless.
- "`using DecryptionParams`" — The file is created based on `EncryptionParams`, not `DecryptionParams`.
- "`See the example script in examples/tensorize_vllm_model.py`" — This could also optionally accept a `DecryptionParams` object. If instead this is only the step meant to parse raw CLI data, and something like a premade `DecryptionParams` wouldn't apply, it is a slightly strange mix to expect reified Python concepts like a callable `filter_func` in the same parameter list.
- "`A relative path within the tensorizer_uri directory can also be used`" — Encryption at rest doesn't make sense if  you store the key right next to it. Using this for in-transit encryption also wouldn't make sense since the exact same URL and connection settings would be used to transmit the key as to transmit the file data, so if the connection is secure, you wouldn't need encryption, and if it isn't, you shouldn't be reading it that way. This would only be worthwhile if someone had obtained the key some other way and then moved it into a file in the `tensorizer_uri` directory for this one purpose. This should just be an absolute path (or the OS's default local path handling, relative to the CWD) so that we don't implicitly suggest that people do this.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -33,49 +31,50 @@ def _is_vllm_model(model_config: ModelConfig = None,
+        -> Union[None, str]:
+    if keyfile is not None:
+        # Check if the keyfile is a relative path
+        if keyfile.count('/') == 0:
```
This is not a way to check for a relative path. `./key` and `subfolder/key` are relative paths.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -33,49 +31,50 @@ def _is_vllm_model(model_config: ModelConfig = None,
+        # Check if the keyfile is a relative path
+        if keyfile.count('/') == 0:
+            tensorizer_dir = os.path.dirname(tensorizer_uri)
+            keyfile = os.path.join(tensorizer_dir, keyfile)
```
Joining via `/` is more accurate than `os.path.join` when the components may be URIs rather than filesystem paths

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -33,49 +31,50 @@ def _is_vllm_model(model_config: ModelConfig = None,
-        param.data = param.data.contiguous()
+def _validate_keyfile_path(keyfile: str = None,
+                           tensorizer_uri: str = None) \
+        -> Union[None, str]:
```
- `Union[None, x]` → `Optional[x]`, in general.
- `keyfile` is an `Optional[str]` here.
- `tensorizer_uri` is sort of an `Optional[str]` here, though it's non-optional when the `keyfile` isn't `None`, and that isn't handled anywhere. It's always able to be provided, so it can just have no default value.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -84,68 +83,71 @@ def __post_init__(self):
         self.s3_secret_access_key = (os.environ.get("S3_SECRET_ACCESS_KEY")
```
Are these environment variables documented anywhere in vLLM? This seems like something that should be passed to the args class rather than being generated out of thin air whenever this object happens to be initialized.

sangstar committed:
Run yapf and ruff

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -84,68 +83,71 @@ def __post_init__(self):
+
+        # Create the argument group
+        group = parser.add_argument_group(
+            'Tensorizer Options',
```
Group names are lowercase by convention
```suggestion
            'tensorizer options',
```

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -84,68 +83,71 @@ def __post_init__(self):
+            'Tensorizer Options',
+            description=('Options for configuring the behavior of the '
+                         'TensorDeserializer'
+                         ' when loading tensors from a serialized model.')
```
- `TensorDeserializer` is the code interface, this is the command line interface
- "`when loading tensors from a serialized model.`" — Models are always loaded from *some* serialized state
```suggestion
            description=('Options for configuring the behavior of the'
                         ' tensorizer deserializer when --load-format=tensorizer')
```
This is an example; I don't know if `--load-format` is an actual CLI flag, but whatever enables tensorizer deserialization would go there (some flag, some required filename extension, etc.).

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -84,68 +83,71 @@ def __post_init__(self):
             help="A `DecryptionParams` object holding a password or key to"
-            "use for decryption. ``None`` (the default) means no decryption.",
+                 " use for decryption. ``None`` (the default) means no "
+                 "decryption.",
```
This help text needs to be updated. It also needs to be documented whether or not this accepts network URIs.

Eta0 replied:
Please also remember that reStructuredText and Markdown are not valid for CLI help text, and to adjust formatting accordingly if you are copy-pasting docstrings. Directly copy-pasting docstrings for a programmatic interface into a CLI is often a bad idea in general, though.

Eta0 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -84,68 +83,71 @@ def __post_init__(self):
+            action="store_true",
+            help="If enabled, `tensorizer` will force a HTTP connection to "
+                 "tensorizer-uri, if applicable, instead of HTTPS. This is"
+                 " slightly faster, but less secure.",
```
This needs a HUGE warning up front about how bad of an idea this is.

Eta0 commented on vllm/model_executor/weight_utils.py:
```diff
@@ -20,6 +21,9 @@
                                                      QuantizationConfig)
 
+if os.getenv('TENSORIZER_IGNORE_WARNINGS'):
+    warnings.filterwarnings("ignore")
```
This is not an acceptable way to silence a single warning message! **This disables all warnings for all libraries for the entire program!** It's also not a warning from tensorizer, it's a warning from vLLM code (\<you are here\>), so it shouldn't be prefixed by `TENSORIZER_`. It'd be better as something like `VLLM_SILENCE_PERFORMANCE_WARNINGS`.
```suggestion
class PerformanceWarning(UserWarning):
    def __str__(self):
        return (
            f"{super().__str__()}" 
            " (set the VLLM_SILENCE_PERFORMANCE_WARNINGS"
            " environment variable to hide this)"
        )


if (
    os.getenv("VLLM_SILENCE_PERFORMANCE_WARNINGS", "").lower()
    not in ("", "0", "n", "no", "off", "disable")
):
    warnings.simplefilter("ignore", category=PerformanceWarning)
```

Eta0 commented on vllm/model_executor/weight_utils.py:
```diff
@@ -263,7 +263,8 @@ def hf_model_weights_iterator(
     elif load_format == "tensorizer":
-        logger.warning(
+        tensorizer_args = load_format.params
+        warnings.warn(
```
Use a custom warning category here like `PerformanceWarning` from my previous comment to allow it to be easily filterable.

sangstar committed:
fix: Add specific category for warnings with `PerformanceWarning`

sangstar committed:
chore: Multiple fixes from final code review

Multiple fixes have been made mostly pertaining to arg parsing and
docs fixes from final code review. Some changes include the changing
of `tensorizer_args` to a dynamic attribute for `EngineArgs`, as well
as multiple QoL changes for example script.

sangstar committed:
fix: Add `s3_endpoint` as attr for `TensorizerArgs`

sangstar committed:
chore: Remove `filter_func` from CLI args, some doc fixes

sangstar committed:
chore: Allow env var or CLI arg specification for S3 credentials

sangstar committed:
fix: Disallow using `force_http`

sangstar committed:
chore: Remove unnecessary print statement in example script

sangstar committed:
Run yapf and ruff

sangstar committed:
Run yapf and ruff

Also added test files

sangstar committed:
Run yapf and ruff

sangstar committed:
docs: Update `tensorizer` as a `--load-format` in `engine_args.rst`

sangstar committed:
fix: Restore `tensorizer_args` as instance attr to `EngineArgs`

Brings back `tensorizer_args` as an instance attr to `EngineArgs`
and refactors `EngineArgs.from_cli_args` to correctly parse and
instantiate `EngineArgs` and `TensorizerArgs` while preserving
`tensorizer_args` as an attr.

sangstar committed:
Run yapf and ruff

sangstar committed:
chore: Move testing out of own test folder

sangstar committed:
fix: Add `tensorizer >= 2.8.1` to `requirements-rocm.txt` for CI

sangstar committed:
fix: Add version of `tensorizer` that will pass testing suite

Adds a branched version of `tensorizer` that doesn't init CUDA
so as to not init CUDA before forking during tests

sangstar committed:
chore: Add notice that `requirements-dev` dep can be removed `>2.8.1`

Adds comment that the specific `tensorizer` dep for `requirements-dev`
can be dropped after `tensorizer > 2.8.1` releases

sangstar committed:
fix: Resolve double `HfFileSystem` import

sangstar committed:
style: Run `isort`

liiliiliil committed:
[Bugfix] Fix ipv6 address parsing bug (#3641)

njhill committed:
[BugFix] Fix ipv4 address parsing regression (#3645)

jeejeelee committed:
[Kernel] support non-zero cuda devices  in punica kernels (#3636)

jeejeelee committed:
[Doc]add lora support (#3649)

WoosukKwon committed:
[Misc] Minor fix in KVCache type (#3652)

youkaichao committed:
[Core] remove cupy dependency (#3625)

WoosukKwon committed:
[Bugfix] More faithful implementation of Gemma (#3653)

youkaichao committed:
[Bugfix] [Hotfix] fix nccl library name (#3661)

megha95 committed:
[Model] Add support for DBRX (#3660)

AmadeusChan committed:
[Misc] add the "download-dir" option to the latency/throughput benchmarks (#3621)

ywang96 committed:
feat(benchmarks): Add Prefix Caching Benchmark to Serving Benchmark (#3277)

zeppombal committed:
Add support for Cohere's Command-R model (#3433)

Co-authored-by: José Maria Pombal <jose.pombal@unbabel.com>
Co-authored-by: youkaichao <youkaichao@gmail.com>

WoosukKwon committed:
[Docs] Add Command-R to supported models (#3669)

esmeetu committed:
[Model] Fix and clean commandr (#3671)

hxer7963 committed:
[Model] Add support for xverse  (#3610)

Co-authored-by: willhe <hexin@xverse.cn>
Co-authored-by: root <root@localhost.localdomain>

youkaichao committed:
[CI/Build] update default number of jobs and nvcc threads to avoid overloading the system (#3675)

WoosukKwon committed:
[Kernel] Add Triton MoE kernel configs for DBRX on A100 (#3679)

cadedaniel committed:
[Core][Bugfix]Refactor block manager for better testability (#3492)

wenyujin333 committed:
[Model] Add support for Qwen2MoeModel (#3346)

ywang96 committed:
[Kernel] DBRX Triton MoE kernel H100 (#3692)

rkooo567 committed:
[2/N] Chunked prefill data update (#3538)

adamrb committed:
[Bugfix] Update neuron_executor.py to add optional vision_language_config (#3695)

simon-mo committed:
fix benchmark format reporting in buildkite (#3693)

simon-mo committed:
[CI] Add test case to run examples scripts (#3638)

esmeetu committed:
[Core] Support multi-node inference(eager and cuda graph) (#3686)

WoosukKwon committed:
[Kernel] Add MoE Triton kernel configs for A100 40GB (#3700)

WoosukKwon committed:
[Bugfix] Set enable_prefix_caching=True in prefix caching example (#3703)

simon-mo committed:
fix logging msg for block manager (#3701)

youkaichao committed:
[Core] fix del of communicator (#3702)

IKACE committed:
[Benchmark] Change mii to use persistent deployment and support tensor parallel (#3628)

simon-mo committed:
bump version to v0.4.0 (#3705)

youkaichao committed:
Revert "bump version to v0.4.0" (#3708)

rkooo567 committed:
[Test] Make model tests run again and remove --forked from pytest (#3631)

Co-authored-by: Simon Mo <simon.mo@hey.com>

WoosukKwon committed:
[Misc] Minor type annotation fix (#3716)

youkaichao committed:
[Core][Test] move local_rank to the last arg with default value(#3711)

[Core][Test] move local_rank to the last arg with default value to keep api compatible (#3711)

simon-mo committed:
add ccache to docker build image (#3704)

yhu422 committed:
Usage Stats Collection (#2852)

esmeetu committed:
[BugFix] Fix tokenizer out of vocab size (#3685)

esmeetu committed:
[BugFix][Frontend] Fix completion logprobs=0 error (#3731)

ywang96 committed:
[Bugfix] Command-R Max Model Length (#3727)

simon-mo committed:
bump version to v0.4.0 (#3712)

hongxiayang committed:
[ROCm][Bugfix] Fixed several bugs related to rccl path and attention selector logic (#3699)

simon-mo committed:
usage lib get version another way (#3735)

njhill committed:
[BugFix] Use consistent logger everywhere (#3738)

youkaichao committed:
[Core][Bugfix] cache len of tokenizer (#3741)

bnellnm committed:
Fix build when nvtools is missing (#3698)

simon-mo committed:
CMake build elf without PTX (#3739)

mawong-amd committed:
[Kernel] Layernorm performance optimization (#3662)

youkaichao committed:
[Doc] Update installation doc (#3746)

[Doc] Update installation doc for build from source and explain the dependency on torch/cuda version (#3746)
Co-authored-by: Zhuohan Li <zhuohan123@gmail.com>

robertgshaw2-redhat committed:
[CI/Build] Make Marlin Tests Green (#3753)

WoosukKwon committed:
[Misc] Minor fixes in requirements.txt (#3769)

sangstar committed:
Run yapf and ruff

njhill committed:
[Misc] Some minor simplifications to detokenization logic (#3670)

Some simplifications made for clarity.

Also moves detokenization-related functions from tokenizer.py to detokenizer.py.

ywang96 committed:
[Misc] Fix Benchmark TTFT Calculation for Chat Completions (#3768)

cadedaniel committed:
[Speculative decoding 4/9] Lookahead scheduling for speculative decoding (#3250)

Qubitium committed:
[Misc] Add support for new autogptq checkpoint_format (#3689)

Co-authored-by: Robert Shaw <rshaw@neuralmagic.com>

cadedaniel committed:
[Misc] [CI/Build] Speed up block manager CPU-only unit tests ~10x by opting-out of GPU cleanup (#3783)

bigPYJ1151 committed:
[Hardware][Intel] Add CPU inference backend (#3634)

Co-authored-by: Kunshang Ji <kunshang.ji@intel.com>
Co-authored-by: Yuan Zhou <yuan.zhou@intel.com>

bigPYJ1151 committed:
[HotFix] [CI/Build] Minor fix for CPU backend CI (#3787)

A-Mahla committed:
[Frontend][Bugfix] allow using the default middleware with a root path (#3788)

Co-authored-by: A-Mahla <>

ywang96 committed:
[Doc] Fix vLLMEngine Doc Page (#3791)

sangstar committed:
fix: Add `tensorizer` to mock imports

youkaichao committed:
[CI/Build] fix TORCH_CUDA_ARCH_LIST in wheel build (#3801)

leiwen83 committed:
Fix early CUDA init via get_architecture_class_name import (#3770)

Signed-off-by: Lei Wen <wenlei03@qiyi.com>
Co-authored-by: Lei Wen <wenlei03@qiyi.com>

mgoin committed:
[Bugfix] Add `__init__.py` files for `vllm/core/block/` and `vllm/spec_decode/` (#3798)

youkaichao committed:
[CI/Build] 0.4.0.post1, fix sm 7.0/7.5 binary (#3803)

cadedaniel committed:
[Speculative decoding] Adding configuration object for speculative decoding (#3706)

Co-authored-by: Lily Liu <lilyliupku@gmail.com>

njhill committed:
[BugFix] Use different mechanism to get vllm version in `is_cpu()` (#3804)

robertgshaw2-redhat committed:
[Doc] Update README.md (#3806)

michaelfeil committed:
[Doc] Update contribution guidelines for better onboarding (#3819)

sangstar committed:
perf: Add newest `tensorizer` version that will not init CUDA

sangstar committed:
fix: Adjust `tensorizer` version for `requirements-dev.txt`

sangstar committed:
chore: Rebase and fix carrying over changes to `arg_utils` typing

sangstar committed:
fix: Add `tensorizer` to `requirements-cpu.txt`

rkooo567 committed:
[3/N] Refactor scheduler for chunked prefill scheduling (#3550)

AdrianAbeyta committed:
Enable scaled FP8 (e4m3fn) KV cache on ROCm (AMD GPU) (#3290)

Co-authored-by: Gregory Shtrasberg <Gregory.Shtrasberg@amd.com>
Co-authored-by: HaiShaw <hixiao@gmail.com>
Co-authored-by: AdrianAbeyta <Adrian.Abeyta@amd.com>
Co-authored-by: Matthew Wong <Matthew.Wong2@amd.com>
Co-authored-by: root <root@gt-pla-u18-08.pla.dcgpu>
Co-authored-by: mawong-amd <156021403+mawong-amd@users.noreply.github.com>
Co-authored-by: ttbachyinsda <ttbachyinsda@outlook.com>
Co-authored-by: guofangze <guofangze@kuaishou.com>
Co-authored-by: Michael Goin <mgoin64@gmail.com>
Co-authored-by: jacobthebanana <50071502+jacobthebanana@users.noreply.github.com>
Co-authored-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

sangstar committed:
perf: Add concurrent reading to `TensorDeserializer`

sangstar committed:
docs: Add `num_readers` docstring

WoosukKwon committed:
[Misc] Publish 3rd meetup slides (#3835)

sighingnow committed:
[BugFix] Pass tokenizer_config to local_tokenizer_group (#3754)

Signed-off-by: Tao He <sighingnow@gmail.com>

michaelfeil committed:
[Core] Enable hf_transfer by default if available (#3817)

WoosukKwon committed:
[Bugfix] Add kv_scale input parameter to CPU backend (#3840)

mgerstgrasser committed:
[Core] [Frontend] Make detokenization optional (#3749)

Co-authored-by: Nick Hill <nickhill@us.ibm.com>

Commit:
[Bugfix] Fix args in benchmark_serving (#3836)

Co-authored-by: Roger Wang <ywang@roblox.com>

gty111 committed:
[Benchmark] Refactor sample_requests in benchmark_throughput (#3613)

Co-authored-by: Roger Wang <ywang@roblox.com>

sangstar committed:
chore: Replace `PerformanceWarning` after rebase

sangstar committed:
Run yapf and ruff

youkaichao committed:
[Core] manage nccl via a pypi package & upgrade to pt 2.2.1 (#3805)

mgoin committed:
[Hardware][CPU] Update cpu torch to match default of 2.2.1 (#3854)

saurabhdash2512 committed:
[Model] Cohere CommandR+ (#3829)

youkaichao committed:
[Core] improve robustness of pynccl (#3860)

SeanGallen committed:
[Doc]Add asynchronous engine arguments to documentation. (#3810)

Co-authored-by: Simon Mo <simon.mo@hey.com>
Co-authored-by: Roger Wang <136131678+ywang96@users.noreply.github.com>

youkaichao committed:
[CI/Build] refactor dockerfile & fix pip cache

[CI/Build] fix pip cache with vllm_nccl & refactor dockerfile to build wheels (#3859)

cadedaniel committed:
[Misc] Add pytest marker to opt-out of global test cleanup (#3863)

cadedaniel committed:
[Misc] Fix linter issues in examples/fp8/quantizer/quantize.py (#3864)

noamgat committed:
[Bugfix] Fixing requirements.txt (#3865)

WoosukKwon committed:
[Misc] Define common requirements (#3841)

tdoublep committed:
Add option to completion API to truncate prompt tokens (#3144)

rkooo567 committed:
[Chunked Prefill][4/n] Chunked prefill scheduler. (#3853)

Isotr0py committed:
[Bugfix] Fix incorrect output on OLMo models in Tensor Parallelism (#3869)

youkaichao committed:
[CI/Benchmark] add more iteration and use median for robust latency benchmark (#3889)

youkaichao committed:
[Core] enable out-of-tree model register (#3871)

youkaichao committed:
[Core] latency optimization (#3890)

Isotr0py committed:
[Bugfix] Fix Llava inference with Tensor Parallelism. (#3883)

SUDA-HLT-ywfang committed:
[Model] add minicpm (#3893)

egortolmachev committed:
[Bugfix] Added Command-R GPTQ support (#3849)

Co-authored-by: Egor Tolmachev <t333ga@gmail.com>

Ki6an committed:
[Bugfix] Enable Proper `attention_bias` Usage in Llama Model Configuration (#3767)

Co-authored-by: roy <jasonailu87@gmail.com>

mawong-amd committed:
[Hotfix][CI/Build][Kernel] CUDA 11.8 does not support layernorm optimizations (#3782)

esmeetu committed:
[BugFix][Model] Fix commandr RoPE max_position_embeddings (#3919)

youkaichao committed:
[Core] separate distributed_init from worker (#3904)

cadedaniel committed:
[Misc] [Core] Implement RFC "Augment BaseExecutor interfaces to enable hardware-agnostic speculative decoding" (#3837)

jsato8094 committed:
[Bugfix] Fix KeyError on loading GPT-NeoX (#3925)

jpvillam-amd committed:
[ROCm][Hardware][AMD] Use Triton Kernel for default FA on ROCm (#3643)

Co-authored-by: jpvillam <jpvillam@amd.com>
Co-authored-by: Gregory Shtrasberg <Gregory.Shtrasberg@amd.com>
Co-authored-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

jeejeelee committed:
[Misc] Avoid loading incorrect LoRA config (#3777)

zedong-peng committed:
[Benchmark] Add cpu options to bench scripts (#3915)

zhaotyer committed:
[Bugfix]  fix utils.py/merge_dict func TypeError: 'type' object is not subscriptable (#3955)

Co-authored-by: tianyi_zhao <tianyi.zhao@transwarp.io>

huyiwen committed:
[Bugfix] Fix logits processor when prompt_logprobs is not None (#3899)

tjohnson31415 committed:
[Bugfix] handle prompt_logprobs in _apply_min_tokens_penalty (#3876)

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

sangstar committed:
fix: Fix model output on deserialization and add e2e output test

WoosukKwon committed:
[Bugfix][ROCm] Add numba to Dockerfile.rocm (#3962)

jamestwhedbee committed:
[Model][AMD] ROCm support for 256 head dims for Gemma (#3972)

sangstar committed:
fix: Properly ensure test outputs are deterministic, add HF model test

sangstar committed:
fix: Make vLLM tensorizing specification less hacky

sangstar committed:
docs: Add tensorizer link in `engine_args.rst`, docstring to example

Added a link to tensorizer in `engine_args.rst` as well as a docstring
describing `examples/tensorize_vllm_model.py` in more detail.

youkaichao committed:
[Doc] Add doc to state our model support policy (#3948)

Co-authored-by: Roger Wang <136131678+ywang96@users.noreply.github.com>

dmarasco committed:
[Bugfix] Remove key sorting for `guided_json` parameter in OpenAi compatible Server (#3945)

fpaupier committed:
[Doc] Fix getting stared to use publicly available model (#3963)

sangstar committed:
chore: Resolve comments

Resolving comments related to making `num_readers` default set to `1`,
`vllm_tensorized` set to a required arg, and tensorizer no longer
being a mandatory dep.

sangstar committed:
fix: Affirm mandatory `vllm_tensorized` argument change

sangstar committed:
perf: Allow preliminary support deserializing with LoRA adapters

sangstar committed:
fix: Fix requirements.txt passing import tensorizer only if installed

tjohnson31415 committed:
[Bugfix] handle hf_config with architectures == None (#3982)

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>
Co-authored-by: Simon Mo <simon.mo@hey.com>

youkaichao committed:
[Core][Refactor] move parallel_utils into vllm/distributed (#3950)

[WIP][Core][Refactor] move vllm/model_executor/parallel_utils into vllm/distributed and vllm/device_communicators (#3950)

sangstar committed:
fix: Properly ensure import fail if tensorizer not used nor installed

sangstar committed:
perf: Move test location and add testing for LoRA

rkooo567 committed:
[Core][5/N] Fully working chunked prefill e2e (#3884)

youkaichao committed:
[Core][Model] torch.compile for layernorm in commandr (#3985)

[Core][Model] Use torch.compile to accelerate layernorm in commandr (#3985)

rkooo567 committed:
[Test] Add xformer and flash attn tests (#3961)

Co-authored-by: Simon Mo <simon.mo@hey.com>

jikunshang committed:
[Misc] Add indirection layer for custom ops  (#3913)

youkaichao committed:
[Doc][Installation] delete python setup.py develop (#3989)

ywang96 committed:
[Kernel] Fused MoE Config for Mixtral 8x22 (#4002)

sangstar committed:
perf: Add some testing changes, introduce `TensorizerConfig`

sangstar committed:
chore: Add `__init__.py` for `tests/tensorizer`

sangstar committed:
tests: Fix `test_tensorizer.py` to account for new changes

sangstar committed:
tests: Remove `test_tensorizer_api_server.py`

kingljl committed:
punica fix-bgmv-kernel-640 (#4007)

sangstar committed:
Run yapf and ruff; fix tests

sangstar committed:
fix: Revert change to `examples/multilora_inference.py`

bigPYJ1151 committed:
[Hardware][Intel] Isolate CPUModelRunner and ModelRunner for better maintenance (#3824)

Yard1 committed:
[Core] Set `linear_weights` directly on the layer (#3977)

sangstar committed:
Merge remote-tracking branch 'upstream/main' into sangstar/integrate-tensorizer

# Conflicts:
#	requirements-cpu.txt
#	vllm/executor/gpu_executor.py
#	vllm/worker/model_runner.py
#	vllm/worker/worker.py

youkaichao committed:
[Core] init_distributed_environment align with init_process_group(#4014)

[Core][Distributed] make init_distributed_environment compatible with init_process_group (#4014)

sangstar committed:
Merge remote-tracking branch 'upstream/main' into sangstar/integrate-tensorizer

sangstar committed:
perf: Update code to reflect change in #3977

sangstar committed:
chore: Remove accidental syntax error

sangstar committed:
docs: Elaborate on S3 credentialing

sangstar committed:
fix: Properly passing `tensorizer_config` to hf weight loader

sangstar committed:
fix: Fix, test tensorizer uri passing without tensorizer load format

sangstar committed:
docs: Note example script in docs for more information

sangstar committed:
chore: Run yapf and ruff, as well as doc edits

sangstar committed:
fix: Fix `initialize_model_parallel` import

sangstar committed:
tests: Add test for `examples/tensorize_vllm_model.py`

sangstar committed:
tests: Fix lora test

sangstar committed:
Run yapf and ruff

sangstar committed:
fix: Move `tensorize_loader` imports to pass CPU test

sangstar committed:
refactor: Pass `TensorizerArgs` direct to `EngineArgs.add_cli_args`

sangstar committed:
tests: Add api_server test using tensorizer

sangstar committed:
fix: Add `tensorizer_config` to `RayGPUExecutor`

sangstar committed:
tests: Formatting and add test to ensure `tensorizer` load format

sangstar committed:
style: Run yapf on `examples/tensorize_vllm_model.py`

sangstar committed:
style: Run isort on `examples/tensorize_vllm_model.py`

sangstar committed:
style: Fix yapf and isort conflict

sangstar committed:
fix: Remove `tensorizer_args` from `ModelConfig`

sangstar committed:
fix: Add error for device scattering and initial handling for quant

sangstar committed:
perf: Multiple changes in response to comments

Replaced the model initialization process using `LLMEngine`, allowing
vLLM to handle and therefore optimize the initial model loading
process. Added testing for quantization.

sangstar committed:
perf: Final changes to resolve comments

sangstar committed:
fix: Skip tests if cURL not installed, add example script for testing

sangstar committed:
Run yapf and ruff

sangstar committed:
tests: Install cURL for tensorizer tests for testing suite

sangstar committed:
tests: Install libsodium23 for CI tensorizer tests

sangstar committed:
fix: Fix testing import path

sangstar committed:
Run yapf and ruff
